Reword push notifications alert text.

jackrzhang opened the pull request:
This addresses https://github.com/zulip/zulip/issues/9949 and https://github.com/zulip/zulip-mobile/issues/1316.

- Omit the adjective 'new', which is implied.
- Omit the proper noun 'Zulip', which is already shown in the header
of a notification.
- Prefer 'PM' over 'private message'. This abbreviation
is becoming more conventional.
- Prefix stream names with '#'.
- Use the verb 'post' in order to describe the action of a user
sending a message to a topic in a stream.

gnprice commented:
Thanks @jackrzhang ! These look like they should be good changes.

Do you have screenshots of what the notifications look like with these changes? That'd be helpful for making sure they all make sense in context. (These changes only affect iOS -- the Android app uses the info to create its own notifications -- so that's the place to test them.)

jackrzhang commented:
I can produce some screenshots!

rishig commented:
Adding a comment, since I can't figure out how else to get notifications on a PR thread.

fwiw, here is how we do email notification subject lines:
```
{% if show_message_content %}
    {% if group_pm %} Group PMs with {{ huddle_display_name }}
    {% elif mention %} {{ sender_str }} mentioned you
    {% elif private_message %} {{ sender_str }} sent you a message
    {% elif stream_email_notify %} New messages in {{ messages[0].header.plain }}
    {% endif %}
{% else %}
    New missed message{{ message_count|pluralize }}
{% endif %}
{% if realm_name_in_notifications %} in {{ realm_str }}
{% endif %}
```

jackrzhang commented:
<p align="center">Screenshots, taken on my iPhone 7: </p>

<p align="center">
<img alt="img_1277" src="https://user-images.githubusercontent.com/12771126/46119160-a9b5a180-c1be-11e8-98b4-654a94d0e8e8.PNG" width="300" />
</p>

<p align="center">
<img alt="img_1278" src="https://user-images.githubusercontent.com/12771126/46119171-b4703680-c1be-11e8-9426-0ddf9ff730f7.PNG" width="300" />
</p>

<p align="center">
<img alt="img_1278" src="https://user-images.githubusercontent.com/12771126/46119172-b4703680-c1be-11e8-8dc2-dd46c7d07d21.PNG" width="300" />
</p>

<p align="center">
<img alt="img_1277" src="https://user-images.githubusercontent.com/12771126/46119173-b508cd00-c1be-11e8-9c49-3765d15bf28e.PNG" width="300" />
</p>

jackrzhang commented:
Looking at these screenshots now, I'm actually not sure if I prefer 'PM' over 'Private message'. I think I would be ok with either, though.

jackrzhang commented:
Also, @rishig FYI - you should be able to subscribe to notifications on any issue or pull request via the right sidebar:
![image](https://user-images.githubusercontent.com/12771126/46119356-6445a400-c1bf-11e8-9471-85979985d02b.png)

jackrzhang commented:
I think it could be good, for consistency, to make the email notification subject lines either the same or just without abbreviations.

rishig commented:
Chatted offline. Conclusion was to follow gmail's conventions. (Also, that we should in no way copy this for Android notifications, which are considerably more powerful.)

I may be forgetting the details 

PM:
**PM with Zoo Lip**

Group PM:
**Zoo Lip**
**Moo Lip, Too Lip, and Zoo Lip**

Stream:
**Zoo Lip**
**stream name > topic**

@-mention:
**Zoo Lip mentioned you**
[same second line as Group PM or stream]

rishig commented:
Actually, maybe for PM we can also just use the name: so 
**Zoo Lip**
?

And yeah, prefixing with `#` sounds good.

jackrzhang commented:
This is updated!

I reverted the changes from the initial commit, added some logic for determining APNs alert subtitles, and updated the tests in `test_push_notifications.py` accordingly.

<p align="center">Screenshots, taken on my iPhone 7: </p>

<p align="center">
<img alt="img_1288" src="https://user-images.githubusercontent.com/12771126/46516791-7ad3a700-c820-11e8-99b0-532baa3abe43.jpeg" width="300" />
</p>

<p align="center">
<img alt="img_1290" src="https://user-images.githubusercontent.com/12771126/46516803-8e7f0d80-c820-11e8-941a-a7d0892d1531.jpeg" width="300" />
</p>

<p align="center">
<img alt="img_1291" src="https://user-images.githubusercontent.com/12771126/46516807-9048d100-c820-11e8-8098-16d9ed0c2db1.jpeg" width="300" />
</p>

<p align="center">
<img alt="img_1292" src="https://user-images.githubusercontent.com/12771126/46516809-9179fe00-c820-11e8-9940-6548b0cffe2b.jpeg" width="300" />
</p>

jackrzhang commented:
I think what's causing the Circle CI tests to fail - which is unrelated to the commits on this branch - has been fixed in master. See [this thread on chat.zulip.org](https://chat.zulip.org/#narrow/stream/9-issues/topic/documentation.20tests).

timabbott commented:
@jackrzhang you can rebase your branch to get tests to pass.  

@gnprice @rishig are you folks both happy with the new wording?

jackrzhang commented:
(rebased)

rishig commented:
For Group PM, I could imagine putting all three names, so that it stays consistent as Zoo Lip and Iago have a back and forth.

jackrzhang commented:
Yeah, that sounds good! @rishig 

Will update right now.

jackrzhang commented:
I've updated the commit so that all of the names in a Group PM are displayed in alphabetical order.

<p align="center">
<img alt="img_1306" src="https://user-images.githubusercontent.com/12771126/46779512-063fb300-cccd-11e8-830f-178e25698c6b.jpeg" width="300" />
</p>

<p align="center">
<img alt="img_1307" src="https://user-images.githubusercontent.com/12771126/46779513-06d84980-cccd-11e8-9bf8-9499854eb3b2.jpeg" width="300" />
</p>

gnprice commented:
Looks good -- thanks @jackrzhang !

Nit in a commit message:
```
Progresses #9499.
```
I think you mean #9949 :)

---

For the group PM case I think there may be further iteration we can do, but this is definitely an improvement and needn't block on that.

Specifically, in those last screenshots it doesn't feel totally obvious to me how the different names relate to each other, and in particular who's saying those words (vs just hearing them.)

Maybe this?
> Polonius
> to Iago, Zoo Lep

(That doesn't feel as compatible with the idea of making the recipient list look the same throughout the thread, though.)

Or maybe invert the order (maybe do the same for stream messages too, to match)?
> Iago, Polonius, Zoo Lep
> Polonius says:

timabbott commented:
I like having the recipients above the sender for cases other than 1:1 PMs where we need to specify that.  It matches the model we have in other Zulip clients (recipient on top) and I think it provides the context in a more clear fashion (generally I care more about which conversation mentioned me than who sent the message).

rishig commented:
I'm fine with recipients first (for iOS)

jackrzhang commented:
Updated so that (1) group PM recipients or the stream + topic name are above the sender's name, and (2) either " says:" or " mentioned you:" is appended to the sender's name:

<p align="center">
<img width="300" src="https://user-images.githubusercontent.com/12771126/47326737-0b7ef500-d61f-11e8-9981-8f127214eed7.jpeg" />
</p>

<p align="center">
<img width="300" src="https://user-images.githubusercontent.com/12771126/47326840-76c8c700-d61f-11e8-990a-ef183dd5a8bc.jpeg" />
</p>

<p align="center">
<img width="300" src="https://user-images.githubusercontent.com/12771126/47326847-7fb99880-d61f-11e8-88c4-3b2835307c29.jpeg" />
</p>

<p align="center">
<img width="300" src="https://user-images.githubusercontent.com/12771126/47326897-cad3ab80-d61f-11e8-9bd3-c19e5317ffe6.jpg" />
</p>

rishig commented:
We should remove the `says:`; at least on Android I've never seen that, and I'm not sure how it'll interact with i18n.

jackrzhang commented:
An iOS notification for Gmail has the email's sender name as the title, which is currently "Zulip missed messages", the subject line as the subtitle, which is could be either "__ sent you a message" or "__ mentioned you", and then the message's content in the body.

gnprice commented:
For a 1:1 PM I think the "says" is superfluous and not common. I'd take it out, and just use the sender's name, like in the [previous version](https://github.com/zulip/zulip/pull/10477#issuecomment-427246902).

Same for stream messages, I think; so
> **#design > font discussion**
> **Zoo Lep**
> Lorem ipsum dolor sit amet

For group PMs, though, I think having no words at all is a bit confusing:
> **Iago, Polonius, Zoo Lep**
> **Polonius**
> Lorem ipsum dolor sit amet

So I think "Polonius says:" is a good solution. It might also be enough to add a colon:

> **Iago, Polonius, Zoo Lep**
> **Polonius:**
> Lorem ipsum dolor sit amet

I think translation will be fine. Other languages can lean on the typographical distinction to help separate the sender from the content, just like the English does. A lot of languages might literally translate the "X says"; others might go for an equivalent of "from X" where that works better.

(But again, all of these are a clear improvement over the status quo. We should merge any version, and we can always fine-tune thereafter.)

timabbott commented:
I think removing the `says` and just using a `:` in the places where it'd be ambiguous (stream message or group PM) looks pretty readable, so I did that tweak and merged this as f116aba490d764e325614980c781180af2cc9378.  I'll deploy this on chat.zulip.org shortly so that we can get feedback from experiencing this in real life.  

Thanks for all the work on this @jackrzhang!  

Are there other follow-ups we want to do for #9949?

gnprice commented:
Thanks @timabbott ! And thanks again @jackrzhang and @rishig for the wordsmithing :smile: 

I think this resolves #9949. There may be further tweaks we make in the future, but we'll be starting from what feels like a pretty different place from what that issue describes.

rishig commented:
I think if we're using a `:`, then we don't need the newline after it. So
**Iago, Polonius, Zoo Lep**
**Zoo Lep**: Lorem ipsum dolor .. 

instead of 
**Iago, Polonius, Zoo Lep**
**Zoo Lep:**
Lorem ipsum dolor ..

timabbott commented:
That change could be an improvement, but keep in mind that the fields are "title", "subtitle", and "content", and I'm not sure splitting a sentence between the "subtitle" and "content" fields will actually work out well.

jackrzhang commented:
About 2 months ago, I took some notes on how Slack was formatting iOS notifications - 
```
(No titles for messages)

1-1 DMs:
@[sender]: content
@Jack Zhang: Lorem ipsum...

Group DMs:
[List of users] @[sender]: content
Jack Zhang, John Smith, Mary Jane @ Jack Zhang (T.03): Lorem ipsum...

Mentions:
#[channel] @[sender]: @[mention] content
#general @Jack Zhang: @everyone Lorem ipsum... 
#general @Jack Zhang: @John Smith Lorem ipsum...
```

I thought it was notable that the notifications weren't using the bolded titles or subtitles. I like what this PR has merged much more :D
